securedrop_client/api_jobs/downloads.py

```diff
@@ -42,22 +43,44 @@ def __init__(self, data_dir: str) -> None:
         self.data_dir = data_dir
 
     def _get_realistic_timeout(self, size_in_bytes: int) -> int:
-        """
-        Returns a realistic timeout in seconds for a file, message, or reply based on its size.
+        '''
+        Return a realistic timeout in seconds for a file, message, or reply based on its size.
+
+        This simply scales the timeouts per file so that in general it increases as the file size
+        increases.
 
         Note that:
-        * the size of the file provided by server is in bytes (the server computes it using
-        os.stat.ST_SIZE)
-        * A reasonable time to download a 1 MB file download time according to OnionPerf was
-        ~7.5s:
-        https://metrics.torproject.org/torperf.html?start=2019-05-02&end=2019-07-31&server=onion&filesize=1mb
-
-        This simply scales the timeouts per file, so that a small file times out much faster than a
-        large file, which is given a long time to complete.
-        """
-
-        SCALING_FACTOR_FILE_DL_TIME_SECS_PER_MB = 7.5
-        return math.ceil(size_in_bytes * SCALING_FACTOR_FILE_DL_TIME_SECS_PER_MB)
+
+        * The size of the file provided by server is in bytes (the server computes it using
+          os.stat.ST_SIZE).
+
+        * The following times are reasonable estimations for how long it should take to fetch a
+          file over Tor according to Tor metrics given a recent three month period in 2019:
+
+          50 KiB  (51200 bytes)   =  4   seconds  (  12800 bytes/second)
+          1  MiB  (1049000 bytes) =  7.5 seconds  (~139867 bytes/second)
+
+          For more information, see:
+          https://metrics.torproject.org/torperf.html?start=2019-05-02&end=2019-07-31&server=onion
+
+        * As you might expect, this method returns timeouts that are larger than the expected
+          download time, which is why the rates below are slower than what you see above with the
+          Tor metrics, e.g. instead of setting SMALL_FILE_DOWNLOAD_TIMEOUT_BYTES_PER_SECOND to
+          12800 bytes/second, we set it to 10000 bytes per second.
+
+        * Minimum timeout allowed is 3 seconds.
+        '''
+        SMALL_FILE_DOWNLOAD_TIMEOUT_BYTES_PER_SECOND = 10000.0
+        DOWNLOAD_TIMEOUT_BYTES_PER_SECOND = 100000.0
+        ONE_MEGABYTE = 1000000
+        MINIMUM_TIMEOUT = 3
+
+        if size_in_bytes < ONE_MEGABYTE:
+            timeout = math.ceil(size_in_bytes / SMALL_FILE_DOWNLOAD_TIMEOUT_BYTES_PER_SECOND)
+            return max(MINIMUM_TIMEOUT, timeout)
+
+        timeout = math.ceil(size_in_bytes / DOWNLOAD_TIMEOUT_BYTES_PER_SECOND)
```

Comment:
OK tested a bit more carefully today and the timeouts are too small for the >1 MB case - I'm able to download <1 MB files without issue but anything 1 MB or larger was timing out until I increased this. Want to try a bit with 1 MB files and confirm my finding here? If you agree then we can crank these timeouts a bit. We should make sure we have reasonable timeouts for files 1 MB - 500 MB in size as those are the supported large file sizes for SD (obviously 500 MB will be sloooooow so worth doing in the background while working on something else).

Reply:
So currently, this is what we're doing:

```py
    # ceil(file_size / 10000 bytes/sec) is expected for file sizes less than 1MB
    # ceil(file_size / 100000 bytes/sec) is expected for file sizes greater than or equal to 1MB
    # minimum timeout is 3 seconds for files less than 1MB

    assert one_byte_file_timeout == 3
    assert one_KB_file_timeout == 3
    assert fifty_KB_file_timeout == 5
    assert five_hundred_KB_file_timeout == 50
    assert one_MB_file_timeout == 10
    assert five_MB_file_timeout == 50
    assert five_hundred_MB_file_timeout == 5000
    assert GB_file_timeout == 10000
```

This means:
*  1 byte file times out after 3 seconds
* 1KB file times out after 3 seconds
* 50KB file times out after 5 seconds
* 500KB file times out after 50 seconds
* 1MB file times out after 10 seconds
* 5MB file times out after 50 seconds
* 500MB file times out after 5000 seconds (~1.39 hours)
* 1GB file times out after 10000 seconds (~2.78 hours)

I think the 1MB timeout is too aggressive as you pointed out. We don't want to make the timeouts too large, plus we'll probably introduce a ramp-up timeout policy on request retries, so how about two small adjustment: (1) we use two minimum timeouts: 10s for files smaller than 1MB, otherwise 75s, and (2) we multiply the timeout by 1.5.

With the adjustments, we would end up with:
*  1 byte file times out after 10 seconds
* 1KB file times out after 10 seconds
* 50KB file times out after 10 seconds
* 500KB file times out after 75 seconds
* 1MB file times out after 75 seconds
* 5MB file times out after 75 seconds
* 50MB file times out after 750 seconds (12.5 minutes)
* 500MB file times out after 7500 seconds (~2.08 hours)
* 1GB file times out after 10000 seconds (~4.17 hours)

Do these seem like reasonable default timeouts?

Reply:
those adjusted timeouts do sound reasonable to me - were you able to reliably download 1 MB and 5 MB files in 75 seconds? I had no problem with small < 1MB files so those small file timeouts are definitely solid, and 2 hours is about right for 500 MB (that's also the session timeout on the source interface). 500 MB is the largest file we need to consider since the securedrop source interface rejects files larger than that. 

yeah ultimately we really do need to implement range requests support...  it's gonna be painful when a 500 MB download fails in the first 1 MB and we don't find out until after 2 hours (or when it fails after 499 MB and we restart the entire download).

Reply:
One other suggestion to simplify this a bit, instead of using two different speeds and minimums, we just use the same speed and use a larger minimum of 25 seconds to account for smaller files that take longer, so `timeout = ceil((file_size/100000)*1.5)+25`. I'll go ahead and make this change so you can test it and provide feedback.

Reply:
> were you able to reliably download 1 MB and 5 MB files in 75 seconds

yup, and to make sure I was using the faster rate I downloaded a 1.2MB file, which in my manual tests takes less than 12seconds. According to tor stats, it should take 7.5s to download ~1MB. I ended up pushing the suggestion above `timeout = ceil((file_size/100000)*1.5)+25` which increases the timeouts across the board with 1MB set to a 40s timeout.